src/database/postgres/playersavequeue.cpp

```diff
@@ -0,0 +1,317 @@
+		"INSERT INTO player_inventories (player, inv_id, inv_width, inv_name, inv_size) VALUES "
+			"($1, $2::int, $3::int, $4, $5::int)");
+
+	PGUtil::prepareStatement(m_conn, "add_player_inventory_item",
+		"INSERT INTO player_inventory_items (player, inv_id, slot_id, item) VALUES "
+			"($1, $2::int, $3::int, $4)");
+
+	PGUtil::prepareStatement(m_conn, "remove_player_metadata",
+		"DELETE FROM player_metadata WHERE player = $1");
+
+	PGUtil::prepareStatement(m_conn, "save_player_metadata",
+		"INSERT INTO player_metadata (player, attr, value) VALUES ($1, $2, $3)");
+
+}
+
+PlayerSaveQueue::~PlayerSaveQueue(){
+}
+
+
+void PlayerSaveQueue::enqueue(RemotePlayer *player){
+	{
+		MutexAutoLock lock(m_exception_mutex);
+		if (m_async_exception) {
+			std::rethrow_exception(m_async_exception);
+		}
+	}
+
+	MutexAutoLock lock(m_mutex);
+
+	QueuedPlayerData *queued_data = new QueuedPlayerData;
+	queued_data->serialized_player = new SerializedPlayer(player);
+
+	std::string playername = player->getName();
+	PlayerSAO* sao = player->getPlayerSAO();
+
+	std::vector<const InventoryList*> inventory_lists = sao->getInventory()->getLists();
+	for (u16 i = 0; i < inventory_lists.size(); i++) {
+		const InventoryList* list = inventory_lists[i];
+		std::string inv_id = itos(i);
+
+		SerializedInventory* serialized_inv = new SerializedInventory(playername, inv_id, list);
+		queued_data->inventories.push_back(serialized_inv);
+
+
+		for (u32 j = 0; j < list->getSize(); j++) {
+			std::ostringstream os;
+			list->getItem(j).serialize(os);
+			std::string itemStr = os.str();
+			std::string slotId = itos(j);
+
+			SerializedInventoryItem *invItem = new SerializedInventoryItem(
+				playername,
+				inv_id,
+				slotId,
+				itemStr
+			);
+
+			queued_data->inventory_items.push_back(invItem);
+		}
+	}
+
+	const StringMap &attrs = sao->getMeta().getStrings();
+	for (const auto &attr : attrs) {
+
+		SerializedPlayerMetadata *metadata = new SerializedPlayerMetadata(
+			playername,
+			attr.first,
+			attr.second
+		);
+
+		queued_data->metadata.push_back(metadata);
+	}
+
+	queue.push_back(queued_data);
+}
+
+
+void *PlayerSaveQueue::run(){
+	std::vector<QueuedPlayerData*> save_items;
+
+	while (!stopRequested()){
+		std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+
+		{
+			// move items over here
+			MutexAutoLock lock(m_mutex);
+			if (!queue.empty()){
+				save_items.swap(queue);
+			}
+		}
+
+		this->save(&save_items);
+		save_items.clear();
+	}
+
+
+	if (!queue.empty()){
+		//flush at exit
+		MutexAutoLock lock(m_mutex);
+		this->save(&queue);
+	}
+
+	return nullptr;
+}
+
+void PlayerSaveQueue::save(std::vector<QueuedPlayerData*> *queue){
+	try {
+
+		for (QueuedPlayerData* item: *queue){
+			PGUtil::checkResults(PQexec(m_conn, "BEGIN;"));
+			savePlayer(item);
+			PGUtil::checkResults(PQexec(m_conn, "COMMIT;"));
+			delete item;
+		}
+
+
+	} catch (std::exception &e) {
+		MutexAutoLock lock(m_exception_mutex);
```

Comment:
An exception would cause a memory leak as `savePlayer` is in charge of freeing memory.